pkg/sidecar_controller/snapshot_controller.go

```diff
@@ -0,0 +1,576 @@
+// If the createtion failed for any reason, the Error status is set accordingly.
+// In alpha version, the controller not retry to create the snapshot after it failed.
+// In the future version, a retry policy will be added.
+
+const pvcKind = "PersistentVolumeClaim"
+const apiGroup = ""
+const snapshotKind = "VolumeSnapshot"
+const snapshotAPIGroup = crdv1.GroupName
+
+const controllerUpdateFailMsg = "snapshot controller failed to update"
+
+const IsDefaultSnapshotClassAnnotation = "snapshot.storage.kubernetes.io/is-default-class"
+
+// syncContent deals with one key off the queue.  It returns false when it's time to quit.
+func (ctrl *csiSnapshotSideCarController) syncContent(content *crdv1.VolumeSnapshotContent) error {
+	klog.V(5).Infof("synchronizing VolumeSnapshotContent[%s]", content.Name)
+
+	if metav1.HasAnnotation(content.ObjectMeta, utils.AnnShouldDelete) {
+		klog.V(5).Infof("syncContent[%s]: check ShouldDelete annotation on content.", content.Name)
+		switch content.Spec.DeletionPolicy {
+		case crdv1.VolumeSnapshotContentRetain:
+			klog.V(4).Infof("VolumeSnapshotContent[%s]: policy is Retain, nothing to do", content.Name)
+
+		case crdv1.VolumeSnapshotContentDelete:
+			klog.V(4).Infof("VolumeSnapshotContent[%s]: policy is Delete", content.Name)
+			ctrl.deleteSnapshotContent(content)
+			klog.V(5).Infof("syncContent: check if we should remove Finalizer for VolumeSnapshotContent[%s]", content.Name)
+			// It is a deletion candidate if DeletionTimestamp is not nil and
+			// VolumeSnapshotContentFinalizer is set.
+			if utils.IsContentDeletionCandidate(content) {
+				// Volume snapshot content is a deletion candidate.
+				// Remove the content finalizer.
+				klog.V(5).Infof("syncContent: Content [%s] is a deletion candidate. Check if it is bound to a snapshot.", content.Name)
+				return ctrl.removeContentFinalizer(content)
+			}
+
+		default:
+			// Unknown VolumeSnapshotDeletionolicy
+			ctrl.eventRecorder.Event(content, v1.EventTypeWarning, "SnapshotUnknownDeletionPolicy", "Volume Snapshot Content has unrecognized deletion policy")
+		}
+		// By default, we use Retain policy if it is not set by users
+		klog.V(4).Infof("VolumeSnapshotContent[%s]: the policy is %s", content.Name, content.Spec.DeletionPolicy)
+
+	} else {
+		var err error
+		if !ctrl.isDriverMatch(content) {
+			klog.V(3).Infof("syncContent: Should not provision snapshot for content %s. Driver does not match", content.Name)
+			return nil
+		}
+		klog.V(5).Infof("syncContent: Call CreateSnapshot for content %s", content.Name)
+		if content.Spec.Source.VolumeHandle != nil && content.Status == nil {
+			if err = ctrl.createSnapshot(content); err != nil {
+				ctrl.updateContentErrorStatusWithEvent(content, v1.EventTypeWarning, "SnapshotCreationFailed", fmt.Sprintf("Failed to create snapshot with error %v", err))
+				return err
+			}
+		} else {
+
+			if err = ctrl.checkandUpdateBoundSnapshotStatus(content); err != nil {
```

Comment:
this function name seems not appropriate? It will call ListSnapshots CSI, and update content status